source/common/grpc/async_client_impl.h

```diff
@@ -0,0 +1,206 @@
+#pragma once
+
+#include "envoy/grpc/async_client.h"
+
+#include "common/common/enum_to_int.h"
+#include "common/common/linked_object.h"
+#include "common/grpc/codec.h"
+#include "common/grpc/common.h"
+#include "common/http/async_client_impl.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/utility.h"
+
+namespace Envoy {
+namespace Grpc {
+
+template <class RequestType, class ResponseType> class AsyncClientStreamImpl;
+
+template <class RequestType, class ResponseType>
+class AsyncClientImpl final : public AsyncClient<RequestType, ResponseType> {
+public:
+  AsyncClientImpl(Upstream::ClusterManager& cm, const std::string& remote_cluster_name)
+      : cm_(cm), remote_cluster_name_(remote_cluster_name) {}
+
+  ~AsyncClientImpl() override { ASSERT(active_streams_.empty()); }
+
+  // Grpc::AsyncClient
+  AsyncClientStream<RequestType>*
+  start(const google::protobuf::MethodDescriptor& service_method,
+        AsyncClientCallbacks<ResponseType>& callbacks,
+        const Optional<std::chrono::milliseconds>& timeout) override {
+    std::unique_ptr<AsyncClientStreamImpl<RequestType, ResponseType>> grpc_stream{
+        new AsyncClientStreamImpl<RequestType, ResponseType>(*this, callbacks)};
+    Http::AsyncClient::Stream* http_stream =
+        cm_.httpAsyncClientForCluster(remote_cluster_name_)
+            .start(*grpc_stream, Optional<std::chrono::milliseconds>(timeout));
+
+    if (http_stream == nullptr) {
+      callbacks.onRemoteClose(Status::GrpcStatus::Unavailable);
+      return nullptr;
+    }
+
+    grpc_stream->set_stream(http_stream);
+
+    Http::MessagePtr message = Common::prepareHeaders(
+        remote_cluster_name_, service_method.service()->full_name(), service_method.name());
+    callbacks.onCreateInitialMetadata(message->headers());
+
+    http_stream->sendHeaders(message->headers(), false);
```

Comment:
I'm not sure if I see that in the current implementation (it can do a `closeLocal()` if `end_stream` is true, but it isn't here). I'll add handling to be defensive though, since the interface allows this.